Relativty_Driver/source/Relativty_HMDDriver.cpp

```diff
@@ -108,125 +97,80 @@ void Relativty::HMDDriver::Deactivate() {
-			m_Pose.qRotation.y = this->quat[2];
-			m_Pose.qRotation.z = this->quat[3];
+			this->new_quaternion_available = false;
+			this->new_vector_available = false;
+		} else if (this->new_quaternion_available) {
+			std::shared_lock quatLock(quatMutex);
+			std::copy(this->quat.begin(), this->quat.end(), reinterpret_cast<float *>(&m_Pose.qRotation));
 
 			vr::VRServerDriverHost()->TrackedDevicePoseUpdated(m_unObjectId, m_Pose, sizeof(vr::DriverPose_t));
-			this->new_quaternion_avaiable = false;
-
-		} else if (this->new_vector_avaiable) {
-
-			m_Pose.vecPosition[0] = this->vector_xyz[0];
-			m_Pose.vecPosition[1] = this->vector_xyz[1];
-			m_Pose.vecPosition[2] = this->vector_xyz[2];
+			this->new_quaternion_available = false;
+		} else if (this->new_vector_available) {
+			std::shared_lock vectorLock(vectorMutex);
+			std::copy(this->vector_xyz.begin(), this->vector_xyz.end(), std::begin(m_Pose.vecPosition));
 
 			vr::VRServerDriverHost()->TrackedDevicePoseUpdated(m_unObjectId, m_Pose, sizeof(vr::DriverPose_t));
-			this->new_vector_avaiable = false;
-
+			this->new_vector_available = false;
 		}
 	}
 	Relativty::ServerDriver::Log("Thread2: successfully stopped\n");
 }
 
 void Relativty::HMDDriver::calibrate_quaternion() {
+	std::lock_guard writeLock(quatMutex);
 	if ((0x01 & GetAsyncKeyState(0x52)) != 0) {
-		qconj[0].store(quat[0]);
-		qconj[1].store(-1 * quat[1]);
-		qconj[2].store(-1 * quat[2]);
-		qconj[3].store(-1 * quat[3]);
+		qconj[0] = this->quat[0];
+		qconj[1] = -1 * this->quat[1];
+		qconj[2] = -1 * this->quat[2];
+		qconj[3] = -1 * this->quat[3];
 	}
-	float qres[4];
-
-	qres[0] = qconj[0] * quat[0] - qconj[1] * quat[1] - qconj[2] * quat[2] - qconj[3] * quat[3];
-	qres[1] = qconj[0] * quat[1] + qconj[1] * quat[0] + qconj[2] * quat[3] - qconj[3] * quat[2];
-	qres[2] = qconj[0] * quat[2] - qconj[1] * quat[3] + qconj[2] * quat[0] + qconj[3] * quat[1];
-	qres[3] = qconj[0] * quat[3] + qconj[1] * quat[2] - qconj[2] * quat[1] + qconj[3] * quat[0];
 
-	this->quat[0] = qres[0];
-	this->quat[1] = qres[1];
-	this->quat[2] = qres[2];
-	this->quat[3] = qres[3];
+	this->quat[0] = qconj[0] * this->quat[0] - qconj[1] * this->quat[1] - qconj[2] * this->quat[2] - qconj[3] * this->quat[3];
+	this->quat[1] = qconj[0] * this->quat[1] + qconj[1] * this->quat[0] + qconj[2] * this->quat[3] - qconj[3] * this->quat[2];
+	this->quat[2] = qconj[0] * this->quat[2] - qconj[1] * this->quat[3] + qconj[2] * this->quat[0] + qconj[3] * this->quat[1];
+	this->quat[3] = qconj[0] * this->quat[3] + qconj[1] * this->quat[2] - qconj[2] * this->quat[1] + qconj[3] * this->quat[0];
 }
 
 void Relativty::HMDDriver::retrieve_device_quaternion_packet_threaded() {
-	uint8_t packet_buffer[64];
-	int16_t quaternion_packet[4];
-	//this struct is for mpu9250 support
-	#pragma pack(push, 1)
+	std::array<uint8_t, 64> packet_buffer;
+	std::array<int16_t, 4> quaternion_packet;
+//this struct is for mpu9250 support
+#pragma pack(push, 1)
 	struct pak {
```

Comment:
this struct corresponds to the packet send by hardware, it *needs* to have this structure
```c++
struct pak {
  uint8_t id;
  float quat[4];
  uint8_t rest[47];
};
```
changing `float[4]` to `std::array<float, 4>` may be fine(can't tell until i test it), but putting it at the end of this struct *is definitely not*

Reply:
Thank you for the review. I've reverted the `float[4]`. Concerning the removal `rest`: if the struct was used in the `hid_read` call, it would indeed not be fine. But on line 167 it is only used as an overlay on the contents of `packet_buffer`, so the trailing data won't matter.

Reply:
fine by me, except the fact that `float[4] quat;` won't compile, it needs to be `float quit[4];`

Reply:
Oops. I was on my work laptop, so I used the web editor and clicked save a little too fast.

Reply:
happens to the best of us